aiosmtpd/smtp.py

```diff
@@ -155,48 +155,60 @@ def push(self, msg):
-                arg = line[i+1:].strip()
-            max_sz = (self.command_size_limits[command]
-                      if self.extended_smtp
-                      else self.command_size_limit)
-            if len(line) > max_sz:
-                yield from self.push('500 Error: line too long')
-                continue
-            if (self._tls_handshake_failed
-                    and command != 'QUIT'):                   # pragma: nossl
-                yield from self.push(
-                    '554 Command refused due to lack of security')
-                continue
-            if (self.require_starttls
-                    and (not self._tls_protocol)
-                    and (command not in ['EHLO', 'STARTTLS', 'QUIT'])):
-                # RFC3207 part 4
-                yield from self.push('530 Must issue a STARTTLS command first')
-                continue
-            method = getattr(self, 'smtp_' + command, None)
-            if not method:
-                yield from self.push(
-                    '500 Error: command "%s" not recognized' % command)
-                continue
-            yield from method(arg)
+            try:
+                # XXX this rstrip may not completely preserve old behavior.
+                line = line.decode('utf-8').rstrip('\r\n')
+                log.info('Data: %r', line)
+                if not line:
+                    yield from self.push('500 Error: bad syntax')
+                    continue
+                i = line.find(' ')
+                if i < 0:
+                    command = line.upper()
+                    arg = None
+                else:
+                    command = line[:i].upper()
+                    arg = line[i+1:].strip()
+                max_sz = (self.command_size_limits[command]
+                          if self.extended_smtp
+                          else self.command_size_limit)
+                if len(line) > max_sz:
+                    yield from self.push('500 Error: line too long')
+                    continue
+                if (self._tls_handshake_failed
+                        and command != 'QUIT'):             # pragma: nossl
+                    yield from self.push(
+                        '554 Command refused due to lack of security')
+                    continue
+                if (self.require_starttls
+                        and (not self._tls_protocol)
+                        and (command not in ['EHLO', 'STARTTLS', 'QUIT'])):
+                    # RFC3207 part 4
+                    yield from self.push(
+                        '530 Must issue a STARTTLS command first')
+                    continue
+                method = getattr(self, 'smtp_' + command, None)
+                if not method:
+                    yield from self.push(
+                        '500 Error: command "%s" not recognized' % command)
+                    continue
+                yield from method(arg)
+            except Exception as e:
```

Comment:
I understand that, and most of code checks says that you must think twice before do that. Althrough asyncio makes such catches to be used more often - that only way to catch unexpected things on task level. There is so wide to take IncompleteReadError in it, other connection problems and other things.
For example, you will inherit your own class which changes command_size_limits logic, by placing object with __contains__ method instead of dictionary. If something goes wrong and you have no logging on loop level (by default there is nothing) you will not recieve any logs or exceptions and nothing will work. If there will be some configuration issue, and you found that only on production configuration - that will make you hate this world for using asyncio.
I`ve added a handler for that to make user define how to handle exceptions there. I use Sentry for that and just placed it in handler, to not intrude with my local code inside.
Im not very happy with this implementation myself, but have nothing better then wide catch for that. May be you can make it better.
I have also idea to make a special handler base class with default implementation of methods. It may contain error_handling with default logger.exception implementation, but that in my queue and will be in one of next PRs.